File: app/src/main/java/org/tvheadend/tvhclient/ui/features/startup/StartupFragment.java
package org.tvheadend.tvhclient.ui.features.startup;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.TextView;

import org.tvheadend.tvhclient.MainApplication;
import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.data.repository.AppRepository;
import org.tvheadend.tvhclient.ui.features.MainActivity;
import org.tvheadend.tvhclient.ui.features.settings.SettingsActivity;
import org.tvheadend.tvhclient.data.service.SyncStateReceiver;
import org.tvheadend.tvhclient.util.menu.MenuUtils;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import timber.log.Timber;

// TODO add nice background image

public class StartupFragment extends Fragment {

    @BindView(R.id.progress_bar)
    ProgressBar progressBar;
    @BindView(R.id.state)
    TextView stateTextView;
    @BindView(R.id.details)
    TextView detailsTextView;
    @BindView(R.id.add_connection_button)
    ImageButton addConnectionButton;
    @BindView(R.id.settings_button)
    ImageButton settingsButton;

    private Unbinder unbinder;
    private AppCompatActivity activity;
    @Inject
    protected AppRepository appRepository;
    @Inject
    protected SharedPreferences sharedPreferences;
    private String stateText;
    private String detailsText;
    private SyncStateReceiver.State state;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.startup_fragment, container, false);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        MainApplication.getComponent().inject(this);
        activity = (AppCompatActivity) getActivity();
        setHasOptionsMenu(true);

        if (savedInstanceState != null) {
            state = ((SyncStateReceiver.State) savedInstanceState.getSerializable("state"));
            stateText = savedInstanceState.getString("stateText");
            detailsText = savedInstanceState.getString("detailsText");
        } else {
            state = SyncStateReceiver.State.IDLE;
            stateText = getString(R.string.initializing);
            detailsText = "";
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putSerializable("state", state);
        outState.putString("stateText", stateTextView.getText().toString());
        outState.putString("detailsText", detailsTextView.getText().toString());
        super.onSaveInstanceState(outState);
    }

    private void handleStartupProcedure() {

        if (appRepository.getConnectionData().getItems().size() == 0) {
            Timber.d("No connection available, showing settings button");
            stateText = getString(R.string.no_connection_available);
            progressBar.setVisibility(View.INVISIBLE);
            addConnectionButton.setVisibility(View.VISIBLE);
            addConnectionButton.setOnClickListener(v -> showSettingsAddNewConnection());

        } else if (appRepository.getConnectionData().getActiveItem() == null) {
            Timber.d("No active connection available, showing settings button");
            stateText = getString(R.string.no_connection_active_advice);
            progressBar.setVisibility(View.INVISIBLE);
            settingsButton.setVisibility(View.VISIBLE);
            settingsButton.setOnClickListener(v -> showConnectionListSettings());

        } else {
            Timber.d("Connection is available and active, showing contents");
            showContentScreen();
        }

        stateTextView.setText(stateText);
        detailsTextView.setText(detailsText);
    }

    @Override
    public void onResume() {
        super.onResume();
        handleStartupProcedure();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.startup_options_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_settings:
                showConnectionListSettings();
                return true;
            case R.id.menu_refresh:
                new MenuUtils(activity).handleMenuReconnectSelection();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void showSettingsAddNewConnection() {
        Intent intent = new Intent(activity, SettingsActivity.class);
        intent.putExtra("setting_type", "add_connection");
        startActivity(intent);
    }

    private void showConnectionListSettings() {
        Intent intent = new Intent(activity, SettingsActivity.class);
        intent.putExtra("setting_type", "list_connections");
        startActivity(intent);
    }

    /**
     * Shows the main fragment like the channel list when the startup is complete.
     * Which fragment shall be shown is determined by a preference.
     * The connection to the server will be established by the network status
     * which is monitored in the base activity which the main activity inherits.
     */
    private void showContentScreen() {
        Intent intent = new Intent(activity, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        activity.startActivity(intent);
        activity.finish();
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/startup/StartupFragment.kt
package org.tvheadend.tvhclient.ui.features.startup

import android.content.Intent
import android.os.Bundle
import android.view.*
import androidx.core.view.forEach
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.afollestad.materialdialogs.MaterialDialog
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.databinding.StartupFragmentBinding
import org.tvheadend.tvhclient.ui.base.BaseViewModel
import org.tvheadend.tvhclient.ui.common.interfaces.HideNavigationDrawerInterface
import org.tvheadend.tvhclient.ui.common.interfaces.LayoutControlInterface
import org.tvheadend.tvhclient.ui.common.interfaces.ToolbarInterface
import org.tvheadend.tvhclient.ui.features.settings.SettingsActivity
import org.tvheadend.tvhclient.util.extensions.gone
import org.tvheadend.tvhclient.util.extensions.visible
import timber.log.Timber

class StartupFragment : Fragment(), HideNavigationDrawerInterface {

    private lateinit var binding: StartupFragmentBinding
    private lateinit var startupViewModel: StartupViewModel
    private lateinit var baseViewModel: BaseViewModel
    private var loadingDone = false
    private var connectionCount: Int = 0
    private var isConnectionActive = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = StartupFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        startupViewModel = ViewModelProvider(requireActivity()).get(StartupViewModel::class.java)
        baseViewModel = ViewModelProvider(requireActivity()).get(BaseViewModel::class.java)

        if (activity is LayoutControlInterface) {
            (activity as LayoutControlInterface).forceSingleScreenLayout()
        }

        if (activity is ToolbarInterface) {
            (activity as ToolbarInterface).setTitle(getString(R.string.status))
        }

        setHasOptionsMenu(true)
        loadingDone = false

        Timber.d("Observing connection status")
        startupViewModel.connectionStatus.observe(viewLifecycleOwner,  { status ->
            Timber.d("Received connection status from view model")

            if (status != null) {
                loadingDone = true
                connectionCount = status.first!!
                isConnectionActive = status.second!!
                Timber.d("connection count is $connectionCount and connection is active is $isConnectionActive")
                showStartupStatus()
            } else {
                Timber.d("Connection count and active connection are still loading")
                binding.startupStatus.visible()
                binding.startupStatus.text = getString(R.string.initializing)
                binding.addConnectionButton.gone()
                binding.listConnectionsButton.gone()
            }
        })
    }

    private fun showStartupStatus() {
        if (loadingDone) {
            if (!isConnectionActive && connectionCount == 0) {
                Timber.d("No connection available, showing settings button")
                binding.startupStatus.visible()
                binding.startupStatus.text = getString(R.string.no_connection_available)
                binding.addConnectionButton.visible()
                binding.addConnectionButton.setOnClickListener { showSettingsAddNewConnection() }
                binding.listConnectionsButton.gone()

            } else if (!isConnectionActive && connectionCount > 0) {
                Timber.d("No active connection available, showing settings button")
                binding.startupStatus.visible()
                binding.startupStatus.text = getString(R.string.no_connection_active_advice)
                binding.addConnectionButton.gone()
                binding.listConnectionsButton.visible()
                binding.listConnectionsButton.setOnClickListener { showConnectionListSettings() }

            } else {
                Timber.d("Connection is available and active, showing contents")
                binding.startupStatus.gone()
                binding.addConnectionButton.gone()
                binding.listConnectionsButton.gone()
                baseViewModel.setStartupComplete(true)
            }
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        menu.forEach { it.isVisible = false }
        // Do not show the reconnect menu in case no connections are available or none is active
        menu.findItem(R.id.menu_settings)?.isVisible = loadingDone
        menu.findItem(R.id.menu_reconnect_to_server)?.isVisible = loadingDone && isConnectionActive
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.startup_options_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_settings -> {
                showMainSettings()
                true
            }
            R.id.menu_reconnect_to_server -> {
                activity?.let { activity ->
                    MaterialDialog(activity).show {
                        title(R.string.reconnect_to_server)
                        message(R.string.restart_and_sync)
                        positiveButton(R.string.reconnect) {
                            Timber.d("Reconnect requested, stopping service and updating active connection to require a full sync")
                            baseViewModel.updateConnectionAndRestartApplication(context)
                        }
                        negativeButton(R.string.cancel)
                    }
                }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun showSettingsAddNewConnection() {
        val intent = Intent(context, SettingsActivity::class.java)
        intent.putExtra("setting_type", "add_connection")
        startActivity(intent)
    }

    private fun showConnectionListSettings() {
        val intent = Intent(context, SettingsActivity::class.java)
        intent.putExtra("setting_type", "list_connections")
        startActivity(intent)
    }

    private fun showMainSettings() {
        val intent = Intent(context, SettingsActivity::class.java)
        startActivity(intent)
    }
}
